Port PML's geometric evaluation / interpolation integration for univariate polynomial (#2449)
flintlib/flint: doc/source/nmod_poly.rst

    modulo the modulus.

    Uses Horner's method iteratively.

.. function:: void nmod_poly_evaluate_geometric_nmod_vec_iter(nn_ptr ys, const nmod_poly_t poly, ulong r, slong n)

    Evaluates ``poly`` at the first ``n`` powers
    of the square of ``r``, writing the output values
    to ``ys``. The value of ``r`` should be reduced
    modulo the modulus.

    Uses Horner's method iteratively.

.. function:: void _nmod_poly_evaluate_geometric_nmod_vec_fast_precomp(nn_ptr vs, nn_srcptr poly, slong plen, const nmod_geometric_progression_t G, slong len)

    Evaluates (``poly``, ``plen``) at the ``len`` values given
    by the precomputed geometric progression ``G``. The value of
    ``len`` should be less than or equal to the precomputation size parameter.

.. function:: void _nmod_poly_evaluate_geometric_nmod_vec_fast(nn_ptr ys, nn_srcptr coeffs, slong len, ulong r, slong n, nmod_t mod)

    Evaluates (``coeffs``, ``len``) at the first ``n`` powers
    of the square of ``r``, writing the output values to ``ys``. 
    The value of ``r`` should be reduced modulo the modulus ``mod``
    and of sufficient multiplicative order such that none of 

>     the powers of ``r`` squared are one.

Reviewer: ```suggestion
    the first `n` powers of `r^2` is one.
```

    Uses fast geometric multipoint evaluation, building a temporary geometric progression precomputation.

.. function:: void nmod_poly_evaluate_geometric_nmod_vec_fast(nn_ptr ys, const nmod_poly_t poly, ulong r, slong n)

    Evaluates ``poly``  at the first ``n`` powers
    of the square of ``r``, writing the output values to ``ys``. 
    The value of ``r`` should be reduced modulo the modulus of the polynomial
    and of sufficient multiplicative order such that none of 

>     the powers of ``r`` squared are one.

Reviewer: ```suggestion
    the first `n` powers of `r^2` is one.
```

    ``xs``, writing the output values to ``ys``. The values in
    ``xs`` should be reduced modulo the modulus.

.. function:: void _nmod_poly_evaluate_geometric_nmod_vec_iter(nn_ptr ys, nn_srcptr coeffs, slong len, ulong r, slong n, nmod_t mod)

    Evaluates (``coeffs``, ``len``) at the first ``n`` powers
    of the square of ``r``, writing the output values
    to ``ys``. The value of ``r`` should be reduced
    modulo the modulus.

    Uses Horner's method iteratively.

> .. function:: nmod_poly_evaluate_geometric_nmod_vec_iter(nn_ptr ys, const nmod_poly_t poly, ulong r, slong n)

Reviewer: ```suggestion
.. function:: void nmod_poly_evaluate_geometric_nmod_vec_iter(nn_ptr ys, const nmod_poly_t poly, ulong r, slong n)
```

    Forms the interpolating polynomial using a naive implementation
    of the barycentric form of Lagrange interpolation.

.. function:: void nmod_poly_interpolate_geometric_nmod_vec_fast(nmod_poly_t poly, ulong r, nn_srcptr ys, slong n)
    
    Sets ``poly`` to the unique polynomial of length at most ``n``
    that interpolates the first ``n`` powers of ``r`` and
    values ``ys``. If the interpolating polynomial is shorter than

>     length ``n``, the leading coefficients are set to zero.

Reviewer: ```suggestion
    values ``ys``.
```
(because I don't think this needs to be made explicit, this is what all functions do anyway, they ensure there are no leading zero, or said otherwise, they apply some `normalise` if needed)

    product is not computed.

Geometric progression
--------------------------------------------------------------------------------

.. function:: void nmod_geometric_progression_init(nmod_geometric_progression_t G, ulong r, slong len, nmod_t mod)

    Builds a geometric progression multipoint evaluation / interpolation structure.

    The value of ``r`` should be reduced modulo the modulus ``mod``
    and of sufficient multiplicative order such that none of 
    the powers `1, r^2, r^4, \ldots, r^{2(len-1)}` is one.

    The value of ``len`` should be greater or equal to the degree of the polynomials one

>     whishes to evaluate / interpolate plus one and greater than the number of points one whises to use.

Reviewer: ```suggestion
    The value of ``len`` should be both greater than or equal to the number of evaluation points to be
    considered, and greater than or equal to the length of the polynomials to be evaluated / interpolated.
```

    product is not computed.

Geometric progression
--------------------------------------------------------------------------------

.. function:: void nmod_geometric_progression_init(nmod_geometric_progression_t G, ulong r, slong d, nmod_t mod)

    Builds a geometric progression multipoint evaluation / interpolation structure.

    The value of ``r`` should be reduced modulo the modulus ``mod``
    and of sufficient multiplicative order such that none of 
    the powers of ``r`` squared up to ``d`` are one.

>     The value of ``d`` should be greater or equal to the degree of the polynomials one

Author: Should be mostly tackled by 2772557a5, do tell me if the phrasing is still misleading.

>     whishes to evaluate / interpolate plus one and greater than the number of points one whises to use.

Reviewer: ```suggestion
    wishes to evaluate / interpolate plus one and greater than the number of points one wishes to use.
```
(some typo fixes, but still remains the point about strict / non-strict, this says both "less than or equal" and then "plus one", so this is confusing)

> 

Reviewer: ```suggestion
    This allocates vectors and polynomials for a total space of `8 d - 1` coefficients.

```

Author: I'll remember to write this sort of memory allocation warnings, thanks !
